improve path expansion tutorial (#450)
catalystneuro/neuroconv: docs/user_guide/expand_path.rst

Path Expansion
===============

> Source data often follows a file and folder organizational structure where some of the metadata of the experiment,
> such as the subject and session id, are included in the filepath of the data files. The

Reviewer: ```suggestion
Source data are often stored in an organizational structure where some of the metadata of the experiment,
such as the subject and session IDs, are included in the folder names and file paths of the data. The
```

:py:mod:`~neuroconv.tools.path_expansion` module allows a user to specify this path pattern so that NeuroConv can
identify all matching data and automatically extract the relevant metadata.

Local Path Expander
-------------------
Use the :py:class:`~neuroconv.tools.path_expansion.LocalPathExpander` to find matching paths in your local filesystem.
This class requires a source data specification, in which you can list multiple data interfaces by name, and for each

> provide a ``base_directory`` path and a ``"file_path"`` or ``"folder_path"`` argument in f-string format. The path

Reviewer: ```suggestion
provide a ``base_directory`` path and a "``file_path``" or "``folder_path``" argument in f-string format. The path
```

    from neuroconv.tools.path_expansion import LocalPathExpander

    # Specify source data
    source_data_spec = {
        "spikeglx": {
            "base_directory": "/path/to/raw_data",
            "file_path": "{subject_id}/{session_id}/{session_id}_g0_imec0/{session_id}_g0_imec0.ap.bin"
        },
        "phy" : {
            "base_directory": "/path/to/processed_data"
            "folder_path": "{subject_id}/{session_id}/phy"
        }
    }

    # Instantiate LocalPathExpander
    path_expander = LocalPathExpander()

    # Expand paths and extract metadata
    metadata_list = path_expander.expand_paths(source_data_spec)

    # Print the results
    for metadata in metadata_list:
        print(metadata)

The ``expand_paths`` method returns a list of :py:class:`~neuroconv.utils.dict.DeepDict` objects that contain two

> dictionaries: ``source_data`` and ``metadata``. The ``source_data`` dictionary contains the resolved filepath of each

Reviewer: ```suggestion
dictionaries: ``source_data`` and ``metadata``. The ``source_data`` dictionary contains the resolved path of each
```

``subject_id``, ``session_id``, and ``session_start_time`` are supported, but this approach could in principle be
extended to support extraction of more metadata.

Specifying Metadata Format
--------------------------
The f-string format allows you to constrain the search for more precise metadata matching using the
`Format Specification Mini-Language`_ and the `1989 C standard format codes`_ for datetimes. Below are some common
examples.

Length
~~~~~~
For example, you might have a data path where the ``subject_id`` and ``session_id`` are next to each other, and they
can be disambiguated because the ``subject_id`` is always 4 characters and the ``session_id`` is always 5 characters.
This can be expressed as ``"{subject_id:4}{session_id:5}"``.

Character type
~~~~~~~~~~~~~~
If the ``subject_id`` is always numeric, you could use ``n``, e.g. ``"{subject_id:n}"``, which will match only digits.

Datetimes
~~~~~~~~~
If your session start time is present in your data path, you can indicate this following the
`1989 C standard format codes`_ for datetimes. For example, ``"{session_start_time:%Y-%m-%d}"`` will match
``"2021-01-02"`` and evaluate it to ``datetime.datetime(2021, 1, 2)``.

> 

Reviewer: ```suggestion

Real-world Examples
-----------------------

Below are some full examples of how this feature can be used on some organizational patterns taken from real datasets.

Example 1
~~~~~~~

TODO
```

Author: let's leave this for later once we have some real-world examples to provide

Reviewer: This is actually the main thing I was requesting from this PR; the other subsections on special capabilities are nice, but what we need for the GUIDE docs is some concrete examples of how to use this in practice

Here are some real-world examples

Example 1) [Allen Institute Visual Coding](https://registry.opendata.aws/allen-brain-observatory/)

Nice and simple as a first example

Example file: `/allen-brain-observatory/visual-coding-2p/ophys_movies/ophys_experiment_645691416.h5`

f-string for all such files: `/allen-brain-observatory/visual-coding-2p/ophys_movies/ophys_experiment_{session_id}.h5`

Example 2) [YutaS of Buszaki](https://app.globus.org/file-manager?origin_id=188a6110-96db-11eb-b7a9-f57b2d55370d&origin_path=%2FSenzaiY%2FYMV01%2FYMV01_170818%2F)

Example file: `/SenzaiY/YMV01/YMV01_170818/YMV01_170818.dat`

f-string for all such files: `/SenzaiY/{subject_id}/{subject_id}_{session_start_time:%y%m%d}/{subject_id}_{session_start_time:%y%m%d}.dat`

Example 3) Was thinking to use Froemke's from figshare, but the URL (even the DOI) in this paper no longer lead anywhere? https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7677212/

Do you know how to find that dataset again?

Author: ok, this is fine. Want to add it to this PR in rst?

Reviewer: > Want to add it to this PR in rst?

If the suggestion is accepted, then I'm fine leaving it to you to decide how best to make the section look really nice

Author: @CodyCBakerPhD I've got a lot on my plate. Could you please do it or delegate it to someone else? Otherwise, let's push this PR through as is and format these examples later.

Author: come to think of it, @felixp8 , want to take a stab at this?
